Add EIP-5719: Signature replacing for smart contract wallets (#5719)
ethereum/EIPs: EIPS/eip-5719.md

---
eip: 5719
title: Signature replacement interface
description: Non-interactive replacing of smart contract wallet signatures that became stale due to configuration changes.
author: Agustin Aguilar (@Agusx1211)
discussions-to: https://ethereum-magicians.org/t/erc-signature-replacing-for-smart-contract-wallets/11059
status: Draft
type: Standards Track
category: ERC
created: 2022-09-26
requires: 1271
---

## Abstract

> A proposal for a standard method to re-encode or replace smart contract wallet signatures at any point after the signature was produced. Wallets implementing this EIP can expose a URI that clients can use to replace an invalid signature with a valid one.

Reviewer: These should be complete sentences.

## Motivation

> In contrast to EOA signatures, smart contract wallet signatures can become invalid at any point in time; this poses a challenge to protocols that rely on signatures remaining valid for extended periods of time.

Reviewer: ```suggestion
In contrast to EOA signatures, [EIP-1271](./eip-1271.md) signatures are not necessarily idempotent; they can become invalid at any point in time. This poses a challenge to protocols that rely on signatures remaining valid for extended periods of time.
```

A signature may need to be mutated due to one of the following scenarios:

> 1) The wallet removes a signer that contributed to signing the initial message.
> 2) The wallet uses a Merkle tree to store signers, adding a new signer.
> 3) The wallet uses a Merkle tree to store signatures, adding new signatures.
> 4) The wallet is updated to a new implementation, and the signature schema changes.

Reviewer: ```suggestion
1. The wallet removes a signer that contributed to signing the initial message.
2. The wallet uses a Merkle tree to store signers, adding a new signer.
3. The wallet uses a Merkle tree to store signatures, adding new signatures.
4. The wallet is updated to a new implementation, and the signature schema changes.
```
Use markdown ordered lists

> Signature replacement should be possible in a non-interactive way, since the wallet that originally signed the message may not be available when the signature needs to be validated (E.g. the settlement of a trade in an exchange that uses an off-chain orderbook).

Reviewer: ```suggestion
Non-interactive signature replacement should be possible, since the wallet that originally signed the message may not be available when the signature needs to be validated. An example use-case is the settlement of a trade in an exchange that uses an off-chain order book.
```

> ## Specification

Reviewer: Do you not want to use RFC 8174?

Reviewer 2: I guess it isn't technically required to include this notice, so I'll still approve this PR, but I would ask the authors to include it since the body of the EIP seems to use `MUST`.

```suggestion
## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
```

> The wallet contract must implement the following interface:

Reviewer: ```suggestion
The wallet contract must implement the following function:
```

        },
        "signature": {
            "type": "string",
            "description": "The alternative signature for the given digest."
        }
    }
}
```

### Client process for replacing a signature

A client is an entity that holds a signature and intends to validate it, either for off-chain or on-chain use. To use the smart contract wallet signature, the client must perform the following actions:

1) Try validating the signature using [EIP-1271](./eip-1271.md); if the signature is valid, then the signature can be used as-is.
2) If the signature is not valid, call `getAlternativeSignature(_digest)`, passing the `digest` corresponding to the old signature.
3) If the call fails, no URI is returned, or the content of the URI is not valid, then the signature must be considered invalid.
4) Try validating the new signature using [EIP-1271](./eip-1271.md); if the signature is valid, it can be used as a drop-in replacement of the original signature.
5) If the validation fails, repeat the process from step (2) (notice: if the URI returns the same signature, the signature must be considered invalid).

> > Notice: Clients MUST implement a retry limit when fetching alternative signatures. This limit is up to the client to define.

Reviewer: ```suggestion
Clients MUST implement a retry limit when fetching alternative signatures. This limit is up to the client to define.
```

> 

Reviewer 2: ```suggestion
```

---
eip: 5719
title: Signature replacement interface
description: Non-interactive replacing of smart contract wallet signatures that became stale due to configuration changes.
author: Agustin Aguilar (@Agusx1211)
discussions-to: https://ethereum-magicians.org/t/erc-signature-replacing-for-smart-contract-wallets/11059
status: Draft
type: Standards Track

> category: ERC

Reviewer 2: If this proposal is modifying the RPC interface wallets expose, it should be in the `Interface` category.

Author: It doesn't affect the RPC interface.

A client is an entity that holds a signature and intends to validate it, either for off-chain or on-chain use. To use the smart contract wallet signature, the client must perform the following actions:

1) Try validating the signature using `EIP-1271`; if the signature is valid, then the signature can be used as-is.
2) If the signature is not valid, call `getAlternativeSignature(_digest)`, passing the `digest` corresponding to the old signature.
3) If the call fails, no URI is returned, or the content of the URI is not valid, then the signature must be considered invalid.
4) Try validating the new signature using `EIP-1271`; if the signature is valid, it can be used as a drop-in replacement of the original signature.
5) If the validation fails, repeat the process from step (2) (notice: if the URI returns the same signature, the signature must be considered invalid).

> Notice: Clients MUST implement a retry limit when fetching alternative signatures. This limit is up to the client to define.

## Rationale

The EIP presents a non-interactive mechanism to replace signatures without requiring the user to reconnect the wallet to the client that holds the stale signature.

> This makes the EIP useful in scenarios where the wallet may not be available to re-sign a transaction, for example, when a 3rd party executes a trade on an NFT marketplace.

Reviewer 2: These two paragraphs are more motivation than rationale. The rationale section should explain technical choices made within the EIP, while the motivation explains why the EIP as a whole should exist.

The following standard allows smart contract wallets to expose a URI that clients can use to replace a stale signature with a valid one.

## Motivation

In contrast to EOA signatures, [EIP-1271](./eip-1271.md) signatures are not necessarily idempotent; they can become invalid at any point in time. This poses a challenge to protocols that rely on signatures remaining valid for extended periods of time.

A signature MAY need to be mutated due to one of the following scenarios:

1. The wallet removes a signer that contributed to signing the initial message.
2. The wallet uses a Merkle tree to store signers, adding a new signer.
3. The wallet uses a Merkle tree to store signatures, adding new signatures.
4. The wallet is updated to a new implementation, and the signature schema changes.

Non-interactive signature replacement SHOULD be possible, since the wallet that originally signed the message MAY NOT be available when the signature needs to be validated. An example use-case is the settlement of a trade in an exchange that uses an off-chain order book.

## Specification

The wallet contract MUST implement the following function:

```solidity

> function getAlternativeSignature(bytes32 _digest) external view returns (string);

Reviewer 3: I must be fundamentally missing something reading EIP-5719. 
What type of signature do you refer to? if it's ECDSA, what was the reason we are returning type `string`?
